### PR TITLE
itest: trustless swap based on vPSBT/PSBT sighash flags

### LINKED ISSUE #577: tapgarden/rpc: expose custom sighash flags for vPSBT funding to enable trustless asset swaps
### Background 

Today, we have the [vPSBT API](https://lightning.engineering/posts/2023-06-14-virtual-psbt/) which allows users to do custom spends using assets the daemon owns. Those custom spends can also have their own scripts, and also eventually custom sighash flags as well. Due to the unique nature of the TAP overlay, flags that normally are considered useless, like `SIGHASH_NONE`, are actually of use for _non-interactive_ swap scenarios. In this case, a seller can publish some data, and the "first" (subject to fee bidding, etc) buyer to add enough funds is able to claim those assets. 

#### The "One Little Trick" 🪄

The missing component to enable this today is allowing users to sign on the _TAP_ level with custom sighash flags. We also want to expose users bringing their own PSBT as well. Once this is exposed, `SIGHASH_NONE` can be used to sign "blank checks" (on the TAP layer). However, these blank checks cannot be cashed without also spending the input that encumbers the check input. That input can only be spent by completing a Bitcoin input/output "puzzle", by adding enough extra inputs to pay the seller and make the check claim valid. This process is non-custodial, and non-interactive 😎. 

### Sally The Seller's Signed Offer

Here's an example flow to illustrate the concept: 
  1. Sally decides to part with her 1-of-1 holographic beefzard card gifted to her by her grandpa so she can make a down payment on a condo in the new SpaceX Luna base. 
  2. She has an input `X`, that commits to the beefzard card (and nothing else for simplicity of this example). 
  3. She makes a PSBT (Bitcoin layer) that adds input `X`, and then an output with a bip 86 addr (doesn't commit to any assets). This 1-in-1 out input is signed with `SIGHASH_SINGLE | SIGHASH_ANYONE_CAN_PAY`. The output asks for `1 BTC` on top of the input amount.
       * At this point, the transaction isn't valid. The transaction is missing that 1 BTC input. 
  4. Sally now makes a vPSBT for the output that the buyer will use to claim the beefzard card. This vSPBT output is "loose", in that everything known is specified, but it doesn't commit to a script key, nor internal key. 
  5. Sally _signs_ (or w/e valid witness for the beefzard card) that vOutput, with `SIGHASH_NONE`. This means she doesn't commit to the script key the asset is to be sent to. This is where some of the magic happens. 
  6. Sally publishes her signed PSBT, and also the signed vPSBT. 
       * The only way someone can claim this is to:
            * Provide enough input to pay Sally. 
            * Anchor the vPSBT output into a new output under the control of the buyer. 

### Claiming the Swap Offer On-Chain

For someone to claim this:
   1. They see the offer from Sally on Nostr, or w/e.
   2. They add in their input on the Bitcoin level to complete the transaction. 
   3. They take the vOut, and anchor it in their new output, providing their own script key and internal key
   4. Broadcast, ???, profit.
      * Here ??? means they may need to compete in the mempool with the other would-be buyers. 

### Interactive & Non-Interactive Auctions 

To eliminate the ???, Alice can set up an open auction. She suspends her beefzard in a timelocked output (only needs to be on the TAP layer), then she has others send her offers, which come in the form of the 1-in-1 out described above. She can run things like a dutch auction using the `sequence` field of the execution transaction: as the price descends, so does the sequence field. Initially, only the highest price is available, then as time goes on, the lower prices are us available and unencumbered by the relative time lock.

You can make the process non-interactive again by publishing these time locked orders as outlined above.

### PR DESCRIPTION
**Pull request by GeorgeTsagk**
## Description

Based on #787 , #779 

Closes #577 

This PR introduces a basic itest that follows the flow of a trustless swap that is possible by correctly manipulating the sighash flags on the asset and bitcoin levels. We create an asymmetry between the vPSBT that describes the asset transfer and the PSBT that anchors it in a bitcoin transaction. The vPSBT only commits to the inputs, allowing anyone to define their own output, while the PSBT commits to the output (which asks for more bitcoin than what is provided in the anchor input) and allows someone to bring their own input to fulfill the transaction.

This way, the only way to "claim" the loose assets on the vPSBT layer is by bringing your own bitcoin on the PSBT layer, which would effectively be the pay-out for the asset holder who created this swap offer.

## Todos

- [x] Add swap redemption part
~~- [ ] Add malicious actor test coverage~~

### CODE REVIEW

File: itest/psbt_test.go
Comment (ffranr) [inline itest/psbt_test.go:1766]: Maybe: "This input is committed to the first output" rather than "This input
	// will only commit to itself and Alice's output"
	
I think that we "commit an input to something" is correct rather than "the input commits itself"? I'm not sure how "commit" is commonly used in this context.

File: itest/psbt_test.go
Comment (ffranr) [inline itest/psbt_test.go:1788]: Is this still a "funded" PSBT? I think we've just stripped that funding. I think at this point it's funded with the taproot-asset anchoring only, is that correct?

File: itest/psbt_test.go
Comment (ffranr) [inline itest/psbt_test.go:1846]: Would be good to have this as a method on the universe server.

File: itest/psbt_test.go
Comment (ffranr) [inline itest/psbt_test.go:1867]: What does tidying up mean here?

File: itest/psbt_test.go
Comment (ffranr) [inline itest/psbt_test.go:1922]: Might be worth naming these indexes, something like: `bobPsbtInFundingIdx := uint32(1)`

**Comment by ffranr** [comment]
As requested, I've only reviewed the itest in the last commit.

File: itest/psbt_test.go
Comment (GeorgeTsagk) [reply itest/psbt_test.go:1691]: `txscript.SigHashNone` is used so that we can sign the input and allow its spending while also committing to none of the outputs, allowing somebody (Bob) later to edit the outputs.

If we isolate on the vPSBT level, Alice is basically saying "i'm signing these assets, and anyone can define the output"

File: itest/psbt_test.go
Comment (GeorgeTsagk) [reply itest/psbt_test.go:1671]: `txscript.SigHashNone` commits to all inputs and none of the outputs

File: itest/psbt_test.go
Comment (GeorgeTsagk) [reply itest/psbt_test.go:1737]: The derivation/key related additions that follow these 2 lines are basically required for `TxOut[0]` to actually pay to `alicePkScript`. I decided to keep the first 2 lines together as they indeed express alice's "terms".

Perhaps worth adding a comment as to what these extra lines do.

File: itest/psbt_test.go
Comment (GeorgeTsagk) [reply itest/psbt_test.go:1751]: Yes, this is essentially asking for a btc change output to be created.

Previously we checked for 2 outputs because one was Alice's btc "reward" (index `0`) and one was the anchor of the assets

Now that Bob is bringing in his bitcoin, he might want to add an explicit change at index `2`, since after paying alice at index`0` there may be change from his inputs. He could "merge" the btc change with the assets anchor. I believe keeping them separate is more clean and simple

File: itest/psbt_test.go
Comment (GeorgeTsagk) [reply itest/psbt_test.go:1766]: Hmm not sure about the exact phrasing here. But the comment should definitely capture the fact that the input actually commits to "itself", as it is valid only if it's present in the final tx and also at the same tx as the output it was signing.

File: itest/psbt_test.go
Comment (GeorgeTsagk) [reply itest/psbt_test.go:1788]: Right, will rename to something like "Alice BTC PSBT"

File: itest/psbt_test.go
Comment (GeorgeTsagk) [reply itest/psbt_test.go:1867]: Essentially prepares the asset outputs

File: itest/psbt_test.go
Comment (ffranr) [reply itest/psbt_test.go:1676]: Ok thanks. Does that mean that if you set `vPkt.Inputs[0].SighashType` then it applies to all `vPkt.Inputs`? I think we should add a comment mentioning how this line commits to all inputs.

File: itest/psbt_test.go
Comment (ffranr) [reply itest/psbt_test.go:1696]: Thanks for the explanation!

File: itest/psbt_test.go
Comment (ffranr) [reply itest/psbt_test.go:1758]: Thank you for the explanation.

Given what you said, I think the code comment above is incorrect, it says:
```
	// This bitcoin PSBT should have 1 input, which is the anchor of Alice's
	// assets, and 2 outputs that correspond to Alice's anchor that carries the
	// transfered assets, and one output that carries the bitcoin change.
```
but I think it should say:
```
	// This bitcoin PSBT should have 1 input, which is the anchor of Alice's
	// assets, and 2 outputs that correspond to the BTC payment that Alice
        // will claim (index 0), and Alice's anchor that carries the transferred
        // assets (index 1).
```

File: itest/psbt_test.go
Comment (ffranr) [reply itest/psbt_test.go:1773]: I understand it this way: The input itself does not commit to anything, as it is not a signature. Rather, it is the signature generated by Alice that commits to the input.

I think you're also saying that the signature generated by Alice actually commits to more than just the input. It commits to the input and also the output at index 0. And that the final tx is only valid if it includes this input and output.

**Review by ffranr — COMMENTED** [review]
I've replied to the ongoing review discussions. They are all close to being resolved.

File: itest/psbt_test.go
Comment (GeorgeTsagk) [reply itest/psbt_test.go:1676]: > Does that mean that if you set vPkt.Inputs[0].SighashType then it applies to all vPkt.Inputs

Every different `vPkt.Inputs[x]` can carry its own sighash flag. For example, we could have a `vPkt` with 2 inputs where input `0` uses SigHashAll and input `1` uses SigHashNone.

> I think we should add a comment mentioning how this line commits to all inputs.

In the example above the `vPkt` has only one input. Perhaps I'm not getting what you mean?

File: itest/psbt_test.go
Comment (GeorgeTsagk) [reply itest/psbt_test.go:1773]: > I understand it this way: The input itself does not commit to anything, as it is not a signature. Rather, it is the signature generated by Alice that commits to the input.

Right, my phrasing is wrong there, it's the signature that commits not the input.

> I think you're also saying that the signature generated by Alice actually commits to more than just the input. It commits to the input and also the output at index 0. And that the final tx is only valid if it includes this input and output.

Yeah this input's signature will commit to that input and the output at the same index (`0`). The fact that this input & output need to be in the tx is more like stating the obvious, could just strip that part

**Comment by bhandras** [comment]
Is this branch and the added test supposed to work already or needs changes from other branches? Tried running it but fails on my end (checked out this branch).

**Comment by guggero** [comment]
I might have broken things in the base branch (`commit-virtual-psbts`), so a rebase might be needed.

**Comment by guggero** [comment]
> I approve of the itest commit (last commit). Maybe we need to change the base branch such that this PR only contains that one commit?

I already did that, just needs a rebase. But to avoid rebasing too often, I'd say let's just park this one until the dependent PR is merged.

**Review by guggero — APPROVED** [review]
Needs a rebase, otherwise LGTM :tada:

File: itest/psbt_test.go
Comment (ffranr) [inline itest/psbt_test.go:1941]: nit: too many empty lines here perhaps

File: itest/psbt_test.go
Comment (ffranr) [inline itest/psbt_test.go:1909]: nit: "as we're not further editing the bitcoin transaction" -> "as we will not edit the bitcoin transaction further"

File: itest/psbt_test.go
Comment (ffranr) [inline itest/psbt_test.go:1965]: We've got assertion functions to ensure that the transfer was complete. Maybe consider using https://github.com/lightninglabs/taproot-assets/blob/98634ca6d552d8eec4bb16491dc288f869f529f4/itest/assertions.go#L876-L876

**Review by ffranr — APPROVED** [review]
That was clear and made sense to me. Nice work!

File: itest/psbt_test.go
Comment (Roasbeef) [inline itest/psbt_test.go:1650]: For clarify we could maybe have `aliceScriptKey` just be a blank pub key? In that later on Bob will swap this out in order to claim this himself.

File: itest/psbt_test.go
Comment (Roasbeef) [inline itest/psbt_test.go:1746]: Hmm, not sure if I'm following this properly, but it appears that we're anchoring the vPSBT into the transaction above that has 1-input-1-output. Instead, Bob should be the one anchoring the vPSBT (with the blank signed check from Alice) into an output that _he_ solely controls.

Comment (Roasbeef) [reply itest/psbt_test.go:1746]: Actually scratch the above, I see an output index of `1` is provided. However how can we anchor this ourselves if we don't yet know Bob's public key? In other words, Bob should be the one that takes the vPSBT, and anchors it. Alice's role is to give the signed PSBT that spends her input with an asset, and asks for X amt of BTC by providing less total inputs than the output amt.

Comment (Roasbeef) [reply itest/psbt_test.go:1746]: Ok scratch all that, I see this was just a round about way of doing this, and we'll strip things down to just 1 input 1 output. 

Leaving these comments in place just to demonstrate my thought process while reviewing this.

File: itest/psbt_test.go
Comment (Roasbeef) [inline itest/psbt_test.go:1847]: Given this is present, why do we need to manually insert proofs above?

**Review by Roasbeef — APPROVED** [review]
Nice work! 

I think the flow can be simplified a bit (eg: why do we need to do the fund at the start just to strip it away, instead of just using alice's input), but this is working fine e2e as is. 

Main comment is that we should try to use the `InsertProof` method on the Universe instead of using the semi-deprecated `ImportProofs` call. Or do we use that method since the `InsertProof` will only be picked up if a receive address is known, and in this case Bob never makes one as this is a more adhoc swap?

File: itest/addrs_test.go
Comment (guggero) [reply itest/addrs_test.go:663]: The thing is, if the receiver doesn't expect a transfer (e.g. there is no address that corresponds to the output), then `InsertProof` doesn't really do anything other than inserting the proof into the universe. It's not being picked up as an asset that belongs to the node. That's why I asked offline what mental model you had in mind for interactive transfers. Do we need a call that tells a node: "Hey, that proof you have there in your universe, that's actually an asset you can spend"?

File: itest/psbt_test.go
Comment (GeorgeTsagk) [reply itest/psbt_test.go:1965]: Tried using this but didn't seem to work? Will keep current checks for now

File: itest/psbt_test.go
Comment (GeorgeTsagk) [reply itest/psbt_test.go:1650]: We'll need to derive the anchor key too so I guess this is less lines overall?

Will rename the variable to describe intent

**Pull request merged**
